build_tools/benchmarks/collect_compilation_statistics.py

```diff
@@ -130,97 +133,169 @@ def get_module_path(flag_file: TextIO) -> Optional[str]:
+      benchmark_suite_dir)
+  module_map = {}
+  for category, _ in benchmark_suite.list_categories():
+    benchmark_cases = benchmark_suite.filter_benchmarks_for_category(
+        category=category)
+    for benchmark_case in benchmark_cases:
+      if benchmark_case.benchmark_case_dir is None:
+        raise ValueError("benchmark_case_dir can't be None.")
+      benchmark_case_dir = benchmark_case.benchmark_case_dir
+
+      flag_file_path = benchmark_case_dir / BENCHMARK_FLAGFILE
+      with flag_file_path.open("r") as flag_file:
+        module_path = get_module_path(flag_file)
+
+      if module_path is None:
+        raise RuntimeError(
+            f"Can't find the module file in the flagfile: {flag_file_path}")
+      compilation_info = CompilationInfo(
+          model_name=benchmark_case.model_name,
+          model_tags=tuple(benchmark_case.model_tags),
+          model_source=category,
+          target_arch=benchmark_case.target_arch,
+          compile_tags=tuple(benchmark_case.bench_mode))
+      module_map[compilation_info] = (benchmark_case_dir /
+                                      module_path).resolve()
+
+  return module_map
+
+
+def legacy_get_module_map_and_build_log(args: argparse.Namespace):
+  module_map = get_module_map_from_benchmark_suite(
+      args.build_dir / benchmark_config.BENCHMARK_SUITE_REL_PATH)
+  return module_map, args.build_dir / NINJA_BUILD_LOG
+
+
+def alpha_get_module_map_and_build_log(args: argparse.Namespace):
+  module_map = get_module_map_from_generation_config(
+      serialized_gen_config=args.generation_config.open("r"),
+      e2e_test_artifacts_dir=args.e2e_test_artifacts_dir)
+  return module_map, args.build_log
+
+
 def parse_arguments():
   """Returns an argument parser with common options."""
 
-  def check_dir_path(path):
-    path = pathlib.Path(path)
+  def check_dir_path(path_str: str) -> pathlib.Path:
+    path = pathlib.Path(path_str)
     if path.is_dir():
       return path
     else:
-      raise argparse.ArgumentTypeError(path)
+      raise argparse.ArgumentTypeError(f"{path} is not a directory.")
+
+  def check_file_path(path_str: str) -> pathlib.Path:
+    path = pathlib.Path(path_str)
+    if path.is_file():
+      return path
+    else:
+      raise argparse.ArgumentTypeError(f"{path} is not a file.")
 
   parser = argparse.ArgumentParser()
   parser.add_argument("--output",
-                      required=True,
                       type=pathlib.Path,
                       help="Path to output JSON file.")
-  parser.add_argument(
-      "build_dir",
-      metavar="<build-dir>",
-      type=check_dir_path,
-      help="Path to the build directory containing benchmark suites.")
   parser.add_argument("--verbose",
                       action="store_true",
                       help="Print internal information during execution.")
 
+  subparser = parser.add_subparsers(title="tool version", required=True)
```

Comment:
The unfortunate thing with subparsers is that you end up with rules that some flags have to come before the subcommand name and some have to come after. I think that's confusing. I ran into the same thing and used parser parents so that each subcommand had all the options. Then all flags have to be after the subcommand name, which I think is less surprising. I tried for a bit to make it possible for them to come before *or* after, but never got it to work quite right.

See https://github.com/iree-org/iree/blob/68c4ec21de/build_tools/github_actions/runner/gcp/update_instance_groups.py#L237

Reply:
Oh, that's cool! Done.